flow/data_pipeline/data_pipeline.py

```diff
@@ -157,6 +160,39 @@ def update_baseline(s3, baseline_network, baseline_source_id):
                   Body=new_str.getvalue().replace('\r', '').encode())
 
 
+def get_completed_queries(s3, source_id):
+    """Return the deserialized list of completed queries from S3."""
+    try:
+        completed_queries_obj = \
+            s3.get_object(Bucket='circles.data.pipeline', Key='lambda_temp/{}'.format(source_id))['Body']
+        completed_queries = json.loads(completed_queries_obj.read().decode('utf-8'))
+    except ClientError as e:
+        if e.response['Error']['Code'] == 'NoSuchKey':
+            completed_queries = []
+        else:
+            raise
+    return completed_queries
+
+
+def put_completed_queries(s3, completed_queries):
+    """Put all the completed queries lists into S3 as in a serialized json format."""
+    for source_id, completed_queries_list in completed_queries.items():
+        completed_queries_json = json.dumps(completed_queries_list)
+        s3.put_object(Bucket='circles.data.pipeline', Key='lambda_temp/{}'.format(source_id),
+                      Body=completed_queries_json.encode('utf-8'))
+
+
+def get_ready_queries(completed_queries, new_query):
+    """Return queries whose prerequisite queries are completed."""
+    readied_queries = []
+    unfinished_queries = set(prerequisites.keys()) - set(completed_queries)
```

Comment:
apparently in Python3, `dict.keys()` returns a "set-like view" so it can be used as a set. in that case, perhaps maybe we should make `completed_queries` a set?

Reply:
fixed